sql/2021/resource-hints/pages_almanac_by_device_and_redundant_preload_font_downloads.sql

```diff
@@ -0,0 +1,77 @@
+#standardSQL
+
+# helper to create percent fields
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+    ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+# retruns the number of redundant font downloads
+# use https://replit.com/@NitinPasumarthy/almanac-2021-redundant-font-preloads#index.js to run the JS on several test payloads
+CREATE TEMPORARY FUNCTION getUnnecessaryFontDownloadsCount(almanac_string STRING)
+RETURNS INT64 LANGUAGE js AS '''
+try {
+    var almanac = JSON.parse(almanac_string)
+    if (Array.isArray(almanac) || typeof almanac != 'object') return -1;
+    
+    var nodes = almanac["link-nodes"]["nodes"]
+    nodes = typeof nodes == 'string' ? JSON.parse(nodes) : nodes
+
+    var preloadFontRes = nodes.filter(n => n['as'] === "font" && n['rel'] === 'preload')
+    var fontFiles = preloadFontRes.map(r => {
+        // r is of the form, ./fonts/roboto-regular.woff2
+        var filePathParts = r.href.split("/");
+        // return would be of the form, roboto-regular.woff2
+        return filePathParts[filePathParts.length - 1]
+    });
+    
+    // only include sites which preload at least one font file
+    if (fontFiles.length === 0) {
+        return -1;
+    }
+    
+    var fontsWithExt = {} // {"woff2": Set( 'font1', 'font2' ), "ttf": Set( 'font2' )}
+    for(var i = 0; i < fontFiles.length; i++) {
+        var [fontName, ext] = fontFiles[i].split(".")
+        if(ext in fontsWithExt) {
+            fontsWithExt[ext].add(fontName)
+        } else {
+            fontsWithExt[ext] = new Set([fontName])
+        }
+    }
+
+    var redundantDownloads = 0
+    if("woff2" in fontsWithExt) {
+        for(var ext of Object.keys(fontsWithExt) ) {
+            if(ext !== "woff2") {
+                redundantDownloads += [...fontsWithExt["woff2"]].filter(f => fontsWithExt[ext].has(f)).length
```

Comment:
Not sure I understand how this works. How can the `woff2` bucket contain anything other than font files with `woff2` extensions?

Reply:
Sorry for a counterintuitive implementation. Here all we are doing is 
```js
fontsWithExt['woff2'] - fontsWithExt['woff']
fontsWithExt['woff2'] - fontsWithExt['ttf']
...
```
where `fontsWithExt[extension]` is a **set** of file names with that extension. This way we obtain the total number of font downloads where both woff and woff2 are downloaded which is redundant. The hypothesis is that, browsers that support preload, also support woff2. So, preloading woff in those scenarios is a waste.

Reply:
Oh ok, so given `foo.woff` and `foo.woff2`, "foo" will appear in both the `woff` and `woff2` buckets and counted as a redundant download. Assuming that the base name of the file is the same.

In last year's Fonts chapter, they looked at the font metadata itself to see what the canonical family name of the font was. That's more complicated but it may yield more reliable results, for example to handle the same font with different base names and formats.

Reply:
Oh nice. I did not know about this concept. Do you remember which query has this implementation? I did a quick glance in ./sql/2020/resource-hints folder but nothing stood out matching this description.

Reply:
It was the Fonts chapter, so you'd need to look in /sql/2020/fonts. Here's an example of a query that extracted name metadata from the font file itself: https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2020/fonts/04_14.variable_font.sql

Reply:
Thanks Rick. I'm unable to follow how I can use the info from `JSON_EXTRACT(payload, '$._font_details')` to understand if two fonts have the same base name. Also, we only wish to consider `foo.woff` and `foo.woff2` redundant only if they are in link-preload tags.

```json
{
    "table_sizes": {
        "GSUB": 59,
        "OS/2": 65,
        "cmap": 194,
        "glyf": 1895,
        "head": 49,
        "hhea": 29,
        "hmtx": 63,
        "loca": 60,
        "maxp": 31,
        "name": 296,
        "post": 207
    },
    "names": {
        "1": "iconFont",
        "2": "Regular",
        "3": "iconFont",
        "4": "iconFont",
        "5": "Version 1.0",
        "6": "iconFont",
        "10": "Generated by svg2ttf from Fontello project.",
        "11": "http://fontello.com"
    },
    "OS2": {
        "version": 1,
        "xAvgCharWidth": 540,
        "usWeightClass": 400,
        "usWidthClass": 5,
        "fsType": 0,
        "ySubscriptXSize": 813,
        "ySubscriptYSize": 350,
        "ySubscriptXOffset": 0,
        "ySubscriptYOffset": 70,
        "ySuperscriptXSize": 813,
        "ySuperscriptYSize": 350,
        "ySuperscriptXOffset": 0,
        "ySuperscriptYOffset": 240,
        "yStrikeoutSize": 24,
        "yStrikeoutPosition": 129,
        "sFamilyClass": 0,
        "ulUnicodeRange1": 0,
        "ulUnicodeRange2": 0,
        "ulUnicodeRange3": 0,
        "ulUnicodeRange4": 0,
        "achVendID": "PfEd",
        "fsSelection": 64,
        "usFirstCharIndex": 57345,
        "usLastCharIndex": 57372,
        "sTypoAscender": 500,
        "sTypoDescender": 0,
        "sTypoLineGap": 45,
        "usWinAscent": 501,
        "usWinDescent": 4,
        "ulCodePageRange1": 1,
        "ulCodePageRange2": 0,
        "panose": {
            "bFamilyType": 2,
            "bSerifStyle": 0,
            "bWeight": 5,
            "bProportion": 3,
            "bContrast": 0,
            "bStrokeVariation": 0,
            "bArmStyle": 0,
            "bLetterForm": 0,
            "bMidline": 0,
            "bXHeight": 0
        }
    },
    "post": {
        "formatType": 2,
        "italicAngle": 0,
        "underlinePosition": 5,
        "underlineThickness": 0,
        "isFixedPitch": 0,
        "minMemType42": 0,
        "maxMemType42": 0,
        "minMemType1": 0,
        "maxMemType1": 0
    },
    "counts": {
        "num_cmap_codepoints": 28,
        "num_glyphs": 29
    }
}
```

Reply:
In this case the name of the font is `iconFont`, so you could compare that against other fonts, as opposed to comparing them purely by filename.

Reply:
@tunetheweb We are taking care of
> if a font is preloaded but that font is never used at all (even in another format). I believe there's a Lighthouse test for this.
in the query [sql/2021/resource-hints/pages_consoleLog_by_device_unused_preload.sql](https://github.com/HTTPArchive/almanac.httparchive.org/pull/2264/files#diff-6afb590c58532e8c89bc9993f28d8c11da5c033b309e64fe23649a46dd9a9125). This covers images, fonts...etc.

We created a separate query for the duplicate fonts because they are not flagged in the console or in Lighthouse and all fonts are downloaded, making it quite obscure to find. 

Demo: https://kevinfarrugia.github.io/resource-hints/preload-multiple-fonts

Regarding iframes, haven't considered them at all tbh.

Reply:
I will check whether we are able to update the query to use the metadata to get a more accurate result and compare it with the sample_data.

Reply:
I was unable to use the font metadata since we are limited to the data available in the preload link (i.e. the `href` attribute) and will go ahead with the filename assumption for now.